File: server/src/main/java/com/colonelpanic/n64all/server/model/ControllerState.java
package com.colonelpanic.n64all.server.model;

import com.colonelpanic.n64all.server.service.PlayerController;
import com.google.common.collect.Maps;

import java.util.Map;

public class ControllerState implements PlayerController {
    private Map<ControllerProperty, Integer> properties;

    public ControllerState() {
        this.properties = Maps.newConcurrentMap();

        for (ControllerProperty prop : ControllerProperty.values()) {
            setState(prop, prop.getDefaultValue());
        }
    }

    public void setState(ControllerProperty property, int value) {
        if (isValidPropertyValue(property, value)) {
            properties.put(property, value);
        }
    }

    public int getState(ControllerProperty property) {
        return properties.get(property);
    }

    private boolean isValidPropertyValue(ControllerProperty prop, int value) {
        return prop.getMinValue() <= value && prop.getMaxValue() >= value;
    }
}

File: server/src/main/java/com/colonelpanic/n64all/server/model/ControllerState.kt
package com.colonelpanic.n64all.server.model

import com.colonelpanic.n64all.server.service.PlayerController
import com.google.common.collect.Maps
import org.slf4j.LoggerFactory

class ControllerState : PlayerController {
    private val properties: MutableMap<ControllerProperty, Int>

    private fun isValidPropertyValue(prop: ControllerProperty, value: Int): Boolean {
        return prop.minValue <= value && prop.maxValue >= value
    }

    companion object {
        private val logger = LoggerFactory.getLogger(ControllerState::class.java)
    }

    init {
        properties = Maps.newConcurrentMap()
        for (prop in ControllerProperty.values()) {
            setState(prop, prop.defaultValue)
        }
    }

    override fun setState(property: ControllerProperty?, value: Int) {
        if (isValidPropertyValue(property!!, value)) {
            properties[property] = value
        }
    }

    override fun getState(property: ControllerProperty?): Int {
        return properties[property]!!
    }
}